### PR TITLE
capnp: fix installation, build with fpic and build shared libraries

### PR DESCRIPTION
**Pull request by timkpaine**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

`capnproto` is only building static libraries, and these are not actually linkable. This PR adds a small test case to ensure that the libraries can be linked, and also revamps to build both static and shared libraries (based off the `bamtools.rb` recipe). 

xref: https://github.com/Homebrew/homebrew-core/pull/139476 
`/usr/bin/ld: /home/linuxbrew/.linuxbrew/lib/libkj.a(exception.c++.o): relocation R_X86_64_TPOFF32 against _ZN2kj12_GLOBAL__N_1L19threadLocalCallbackE' can not be used when making a shared object; recompile with -fPIC`

### CODE REVIEW

**Comment by timkpaine** [comment]
The test added fails against the current build (you can see in commit https://github.com/Homebrew/homebrew-core/pull/139577/commits/c2aba45bd6e4debd50b6e783c011bd2e14fc8552, `/usr/bin/ld: /home/linuxbrew/.linuxbrew/Cellar/capnp/1.0.0/lib/libkj.a(exception.c++.o): relocation R_X86_64_TPOFF32 against _ZN2kj12_GLOBAL__N_1L19threadLocalCallbackE' can not be used when making a shared object; recompile with -fPIC
`, https://github.com/Homebrew/homebrew-core/actions/runs/5863531662/job/15897259150?pr=139577#step:4:103

File: Formula/capnp.rb [hunk 1]
```diff
@@ -24,10 +24,22 @@ class Capnp < Formula
...
 
   def install
-    mkdir "build" do
-      system "cmake", "..", *std_cmake_args
-      system "make", "install"
-    end
+    # Build shared library
+    system "cmake", "-S", ".", "-B", "build_shared",
+                    "-DBUILD_SHARED_LIBS=ON",
+                    "-DCMAKE_POSITION_INDEPENDENT_CODE=ON",
```
Comment (carlocab) [inline Formula/capnp.rb:30]:
```suggestion
                    "-DCMAKE_POSITION_INDEPENDENT_CODE=ON",
                    "-DCMAKE_INSTALL_RPATH=#{rpath}",
```

File: Formula/capnp.rb [hunk 2]
```diff
@@ -24,10 +24,22 @@ class Capnp < Formula
...
+    system "cmake", "--install", "build_shared"
+
+    # Build static library
+    system "cmake", "-S", ".", "-B", "build_static",
+                    "-DBUILD_SHARED_LIBS=OFF",
+                    "-DCMAKE_POSITION_INDEPENDENT_CODE=ON",
+                    *std_cmake_args
+    system "cmake", "--build", "build_static"
+    lib.install Dir["build_static/src/capnp/*.a"]
+    lib.install Dir["build_static/src/kj/*.a"]
```
Comment (carlocab) [inline Formula/capnp.rb:42]:
```suggestion
    lib.install buildpath.glob("build_static/src/capnp/*.a")
    lib.install buildpath.glob("build_static/src/kj/*.a")
```

**Review by carlocab — COMMENTED** [review]
Thanks; could you squash your commits together and use the commit message
```
capnp: build with fpic and shared libraries
```
(or similar)?

File: Formula/capnp.rb [hunk 3]
```diff
@@ -24,10 +24,23 @@ class Capnp < Formula
...
+
+    # Build shared library
+    system "cmake", "-S", ".", "-B", "build_shared",
+                    "-DBUILD_SHARED_LIBS=ON",
+                    "-DCMAKE_POSITION_INDEPENDENT_CODE=ON",
+                    "-DCMAKE_INSTALL_RPATH=#{rpath}",
+                    *std_cmake_args
+    system "cmake", "--build", "build_shared"
+    lib.install buildpath.glob("build_static/src/capnp/*.{so,dylib}")
+    lib.install buildpath.glob("build_static/src/kj/*.{so,dylib}")
```
Comment (carlocab) [inline Formula/capnp.rb]: Let's not do this. `cmake` relinks binaries at install time, and we don't want to skip that. We should only do `lib.install` for static libraries.

**Review by carlocab — CHANGES_REQUESTED** [review]
Thanks; but we shouldn't switch around the method of the installation of the shared and static libraries. Shared libraries should be installed using `cmake --install`.

Also, you'll need to do `git fetch origin && git rebase origin/master` to pull in the changes from #139592.

File: Formula/capnp.rb (hunk 3 above)
Comment (timkpaine) [reply Formula/capnp.rb]: annoying, I wanted the executable to use the static link. I'll switch around

File: Formula/c/capnp.rb [hunk 4]
```diff
@@ -24,10 +24,25 @@ class Capnp < Formula
...
+
+    # Build static library
+    system "cmake", "-S", ".", "-B", "build_static",
+                    "-DBUILD_SHARED_LIBS=OFF",
+                    "-DCMAKE_POSITION_INDEPENDENT_CODE=ON",
+                    *std_cmake_args
+    system "cmake", "--build", "build_static"
+
+    # override everything via static, will clobber executables
+    # and include but thats the desired behavior
```
Comment (carlocab) [inline Formula/c/capnp.rb:44]:
```suggestion
    # override everything via static, will clobber executables
    # and include but thats the desired behavior
```
I disagree; that kind of defeats the purpose of having shared libraries. Let's just link the executables to the shared libraries.

Comment (timkpaine) [reply Formula/c/capnp.rb:44]: sure, how would you get this command to work:
`    file.write shell_output("#{bin}/capnp id").chomp + ";\n"`

File: Formula/c/capnp.rb [hunk 5]
```diff
@@ -24,10 +24,23 @@ class Capnp < Formula
   depends_on "cmake" => :build
 
   def install
```
Comment (carlocab) [inline Formula/c/capnp.rb:26]: This should fix your `brew linkage` failure:
```suggestion
  depends_on "cmake" => :build
  uses_from_macos "zlib"

  on_linux do
    depends_on "openssl@3"
  end

  def install
```

**Review by carlocab — APPROVED** [review]
Nice work; thanks for your contribution!

**Pull request merged**